APIdefinition.md

```diff
@@ -0,0 +1,96 @@
+user interface (GUI) programs. Any "interface" via which a user can
+reasonably interact with components of the ISIS project should be
+considered part of the "ISIS Project public API" for the purposes
+of [Semantic Versioning](https://semver.org).
+
+Examples are provided below, but are not meant to be exhaustive or
+complete. The above philosophy should guide our thinking in what
+the public API is.
+
+
+## General Guidance
+
+The public API consists of the following.
+
+**Code-level structures**: Public function, member, and class names,
+argument signatures, and their return signatures.  Where "public"
+is defined as those elements that are typically provided for use external
+to a library or module, dependent on the language.
+
+Some examples are:
+
+- C++ public structures are those that can be access by including
+  the distributed header files for libraries.
+- Python public structures are those that do not begin with an underbar
+  ("_"), as described in PEP 8.
+
+
+**User-facing command-line and GUI programs**: program name, argument
+list, argument defaults, configuration or preference names and their
+defaults, and for those programs with structured output files, the
+structure of those output files are analogous to the return signatures
+of code-level structures.
+
+
+This guidance may be more straightforward to understand and apply
+for code-level structures which have more formalized argument and
+return signatures.  User-facing command-line and GUI programs should
+include verbose information in their documentation that describes what
+a user can expect as far as the structure of the output.
+
+
+## More Specific Examples
+
+Adding new arguments to code level structures does not violate the
+existing public API as long as they provide a default value. For
+example changing a function *doThing(int a, double b)* to *doThing(int
+a, double b, double c=1.0)* does not violate the existing API for
+the *doThing* function as code that calls the *doThing* function will
+still compile without modification.
+
+Adding new optional arguments or arguments with default values to
+command line programs does not violate the existing public API. For
+example, adding a new argument called "check" that has a default
+value of "false" to a command line program does not violate the
+program's public API as scripts and commands that were valid
+previously are still valid after the addition.
+
+
+### Unstructured text files
+
+For anything of an inherently unstructured file-type, like .txt;
+there is no guarnatee of its contents; just that it will exist.
+
+
+### CSV output files
+
+Column names and data types are part of the API.
+
+For example, a program that writes out a CSV should clearly describe
+the output columns which include the fieldname, data type, and
+whether non-data-type values can be present (e.g. a column of
+floating point numbers named "minimum" that could contain the string
+"NULL" instead of a floating point value).
```

Comment:
I think this whole document has to be applied retroactively. So, if we change the type of a column in a CSV or allow an "N/A" entry in a numeric field that would be API breaking. I don't think we are immediately compelled to generate this documentation. It would be very good to have, though, and can be something we include in all future app documentation.

@ladoramkershner This may be something good to consider for our documentation audit.

Reply:
The indicated sentence begins with "For example, ... should ..." which was intended to be for example purposes, and does not expect all of the existing code to have this characteristic, but as development moves forward, and these programs and documentation are revisited, they "should" be more explicit.

If there is a specific change you'd like to see here, please be specific, or resolve.